### ISSUE TITLE
Cannot read property 'write' of null. 

### ISSUE THREAD

**Issue by geoffreysamper** [noodlefrenzy/node-amqp10#295]
Some times when my server is under load and the connection is lost then I recveive this exception. When the connection is restored everything works fine

there was an error during sending  TypeError: Cannot read property 'write' of null
    at Object.frames.writeFrame (C:\github\profiling-tracker-api\node_modules\amqp10\lib\frames.js:56:9)
    at Connection.sendFrame (C:\github\profiling-tracker-api\node_modules\amqp10\lib\connection.js:328:10)
    at SenderLink._sendMessage (C:\github\profiling-tracker-api\node_modules\amqp10\lib\sender_link.js:202:29)
    at sendMessage (C:\github\profiling-tracker-api\node_modules\amqp10\lib\sender_link.js:120:28)
    at C:\github\profiling-tracker-api\node_modules\amqp10\lib\sender_link.js:143:5
    at SenderLink.send (C:\github\profiling-tracker-api\node_modules\amqp10\lib\sender_link.js:113:10)
    at C:\github\profiling-tracker-api\lib\eventHub.js:78:27
    at runCallback (timers.js:637:20)
    at tryOnImmediate (timers.js:610:5)
    at processImmediate [as _immediateCallback] (timers.js:582:5)
[2017-02-16 12:58:58] [LOG]   TypeError TypeError: Cannot read property 'write' of null
    at Object.frames.writeFrame (C:\github\profiling-tracker-api\node_modules\amqp10\lib\frames.js:56:9)
    at Connection.sendFrame (C:\github\profiling-tracker-api\node_modules\amqp10\lib\connection.js:328:10)
    at SenderLink._sendMessage (C:\github\profiling-tracker-api\node_modules\amqp10\lib\sender_link.js:202:29)
    at sendMessage (C:\github\profiling-tracker-api\node_modules\amqp10\lib\sender_link.js:120:28)
    at C:\github\profiling-tracker-api\node_modules\amqp10\lib\sender_link.js:143:5
    at SenderLink.send (C:\github\profiling-tracker-api\node_modules\amqp10\lib\sender_link.js:113:10)
    at C:\github\profiling-tracker-api\lib\eventHub.js:78:27
    at runCallback (timers.js:637:20)
    at tryOnImmediate (timers.js:610:5)
    at processImmediate [as _immediateCallback] (timers.js:582:5)

**Comment by jeffmcaffer** [comment]
We are seeing something very similar but with different outcomes.  Once the queues get messed up they detach and reattach.  Sometimes we see traces like the OP.  Other times it is the sendFrame() call in _sendMessage that is that fails as the connection is null.

More fundamentally, we end up in a state where the promises returned from senderLink.send() are not settling. As a result, the calling service just hangs.  

We poked about in the amqp code and tried a fix.  See [this fork](https://github.com/jeffmcaffer/node-amqp10).  I'm not claiming that it is good/works but wanted to get a sense if this is the right direction.

I wanted to create a repro for testing but that would be hard.  While we seem to get a lot of detach/reattach cycles with Azure service bus, it also seems to take a while to for the problem to show up.

**Comment by geoffreysamper** [comment]
How I could simulate my problem is to load test my server and ensure that the cpu is saturated and then try to run to force a close connection.  [repo](https://github.com/geoffreysamper/amqp-issue-multi-client-same-eventhub)

**Comment by mbroadst** [comment]
@jeffmcaffer @geoffreysamper my hunch here is that when detatch/closing the connection (either manually, or forced) we do not clean up some local bookkeeping that would otherwise indicate that we cannot send data ([SenderLink.prototype.canSend](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/sender_link.js#L39)), which would otherwise queue your messages.  I'll try to put together a unit test exhibiting this around lunch time

**Comment by jeffmcaffer** [comment]
Thanks @mbroadst.  Really appreciate you help here.  We are standing up a production service and hitting this a lot.

I noticed a few things that might help figuring this out.

* typically I see a mess of errors like the one that started this issue THEN I see the link detach and attach
* We switched from multiple AMQPClient objects (one per queue) to just one client that is reused across 5 different queues.  That did not seem to change the behavior.
* There may be two (un/semi) related issues here.  The exceptions noted here and a failure to settle promises. The promise issue is that sometimes we successfully get back a promise from the send() call and the promise is never settled. It seems to happen when there are errors but not always but not for all errors.
* The changes in [my fork](https://github.com/jeffmcaffer/node-amqp10), while not necessarily correct, do point to an issue with _sendMessage failing.  That is, if _sendMessage throws, it is unclear how the returned promise gets settled.  In the !canSend() case the we "queue up" the sendMessage() call.  That call can and does throw out at [line 127](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/sender_link.js#L127) (if _sendMessage throws) when called at [line 248](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/sender_link.js#L248)
* I wonder if this is an issue in other places, for example, sending accept and release
* I've been running with a larger credit limit on the sender link and a rate limiter on calls to send and it is better.  Still see errors but fewer lost promises.  FWIW, my credit is 100 and the ratelimiter is set to 100 per second per process.

I've started seeing forceDetach issue as well.  See trace below.  There is no pattern I can see and these occur on the receiver link too.

```
AmqpProtocolError:
   at SenderLink.forceDetach (at SenderLink.forceDetach (D:\home\site\wwwroot\node_modules\amqp10\lib\sender_link.js:283:30): D:\home\site\wwwroot\node_modules\amqp10\lib\sender_link.js: 283)
   at detachHandler (at detachHandler (D:\home\site\wwwroot\node_modules\amqp10\lib\session.js:441:41): D:\home\site\wwwroot\node_modules\amqp10\lib\session.js: 441)
   at Session._unmap (at Session._unmap (D:\home\site\wwwroot\node_modules\amqp10\lib\session.js:442:33): D:\home\site\wwwroot\node_modules\amqp10\lib\session.js: 442)
   at Session._processEndFrame (at Session._processEndFrame (D:\home\site\wwwroot\node_modules\amqp10\lib\session.js:319:8): D:\home\site\wwwroot\node_modules\amqp10\lib\session.js: 319)
   at Session._processFrame (at Session._processFrame (D:\home\site\wwwroot\node_modules\amqp10\lib\session.js:300:53): D:\home\site\wwwroot\node_modules\amqp10\lib\session.js: 300)
   at Connection._processFrameEH (at Connection._processFrameEH (D:\home\site\wwwroot\node_modules\amqp10\lib\session.js:186:49): D:\home\site\wwwroot\node_modules\amqp10\lib\session.js: 186)
```

**Comment by mbroadst** [comment]
@jeffmcaffer my guess is something like this is going on:

  * the connection is broken because of <somereason> 
  * in this case we erroneously do not cleanup some very important book keeping information
  * you attempt to send a message, and it checks if it can currently send (in order to determine whether to push to pendingSends or not):
    * [this code](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/sender_link.js#L44) is checked, and it erroneously returns `true` for `canSend` because its using invalid information about some connection that is now long gone
  * now we go back to the call site of `_sendMessage` which indeed incorrectly attempts to write to the now non-existent connection, and all hell breaks loose

**Comment by jeffmcaffer** [comment]
Agreed that the exception root cause is elsewhere.  

Try/catch may still be applicable because sendMessage can still throw (it does a network call so unless that is guaranteed to not throw...) and there will be nothing to settle the promise.  If you can't send (!canSend) then sendMessage() will be called from _dispatchPendingSends().  That has no ability to settle the promise.  If I get my promise coding right (not a Bluebird person) sendMessage() throwing is handled by the initial .then() infrastructure (converted to a reject).  However, the !canSend path is sendMessage() is just a function that just happened to be created during a call to new Promise.  The invocation from _dispatchPendingSends() is not in the context of the promise infrastructure so a throw there leaves the promise dangling.

Should be testable by having canSend be false and then throwing from sendMessage.  I've not looked at the test suite yet so not sure whats involved in setting that up.

**Comment by mbroadst** [comment]
@jeffmcaffer its perhaps confusingly named, but the `sendMessage` that is pushed to pendingSends actually [this](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/sender_link.js#L121) which includes the logic for what to do wrt callback policy.  `dispatchPendingSends` in this scenario is called when an `attached` event is processed, and therefore we should be covered there in expecting it not to throw.  Unless I'm missing another potential case here, it seems that the likeliest path for an exception is that `canSend` is operating on bad assumptions. 

I'll try to put a test case together here soon, but if you think you can trigger it other interesting ways the unit tests to look at would be [here](https://github.com/noodlefrenzy/node-amqp10/blob/master/test/unit/sender_link.test.js). We have a fairly flexible mock framework for simulating server response frames, etc.

**Comment by jeffmcaffer** [comment]
Will try to carve off some testing time this afternoon/evening.  

I do still think there is a potential problem with sendMessage().  Consider the pared down version below.  The fact that the sendMessage() function is declare in a Promise constructor is immaterial, it has resolve and reject handlers injected.  Now, when the handler() function runs, sendMessage() is called and doSomethingThatCanThrow() runs and throws.  In that case, who ever calls resolve or reject?  If they never get called, the promise given back originally from send() never settles.

```
    var resolve, reject;    
    var sendMessage = function (err) {
      doSomeThingThatCanThrow();
      if (goodThingHappens())
        resolve(state);
      else
        reject();
    };

    if (!self.canSend()) {
      self._pendingSends.push(sendMessage);
      return;
    }
```
Then sometime later the following code runs

```
handler() {
      var sendMessage = this._pendingSends.shift();
      sendMessage(err);
}
```

**Comment by mbroadst** [comment]
@jeffmcaffer right, what I'm saying is that given the connection is setup properly a call to `_sendMessage` _cannot_ throw. We could wrap the subsequent call to the `sendMessage` wrapper with a try/catch, but it would not be solving the problem you guys are running into imho. So, while I agree that it's probably a safer approach in general, it would have only further hidden the actual problem here

**Comment by geoffreysamper** [comment]
@jeffmcaffer  is possible to share the log file. What I cannot understand is that you said that most of the time the issue occures in attaching mode but if you look at the code of the canSend function it the first line says when not attached return false

**Comment by jeffmcaffer** [comment]
@geoffreysamper that logging code was removed once we isolated the issue as not in our project.  I'll look for an old file or we can look at adding it back if it will help.  You can see the calling code [here](https://github.com/Microsoft/ospo-ghcrawler/blob/develop/lib/amqp10Queue.js).  Imagine push() being called many times per second.

More and more I think there are two separate but related issues and it may be confusing things to have them both in the same issue (happy to split the issue if you like)

1. the exception being thrown (original problem)
2. promises being lost (initially thought to be directly related but increasingly looking like a separate thing)

My code snippets above were related to the second issue.  @mbroadst seems to be on a good path for the first problem.  

The point about attaching is very apt wrt the second problem.  canSend === false is indeed the case where it seems more likely that promises will be lost.  That is because the sendMessage() function is executed outside the context of the Promise infrastructure (i.e., in the attach handling) so there is no one to catch the exception and resolve/reject the promise.

sendMessage can throw as _sendMessage() has several explicit throws and eventually calls down to the stream layer potentially throwing along the way.

One other thing that may be relevant here is that we have many (10s) of worker "loops" running concurrently. Each one gets a message from a queue (one of many), does some stuff, pushes (potentially 100+) messages to a queue, waits on all the promises to settle and repeats.  What we see is that these worker die off.  We narrowed it down to the last part of the loop, waiting on the returned send() promises to complete.  When things go south, it seems that the workers die off slowly (over a minute or two) until none are left and the service stops working as there are a configured number of workers.

The lost promise problem happens with just one work loop but it takes longer. That seems like a load thing.  We mitigate some by rate limiting our calls to send() to 100/s/process.  While that seems to help, it does not solve the problem.

**Comment by jeffmcaffer** [comment]
I found a log.  It was pretty verbose so I trimmed it down, put it in a Word doc and annotated with comments so you could tell what's what.  It does not really show much other than the problem happens.

[ampqlog.docx](https://github.com/noodlefrenzy/node-amqp10/files/805792/ampqlog.docx)

**Comment by geoffreysamper** [comment]
@jeffmcaffer question if I would setup the ospo-ghcrawler project and run it on my env would the issue occur on my env? If this is the case I will try to setup it my env and see better understand the problem.
Is it correct that the client disconnect after a couple of minutes and you can recover from it? The reason that I asked is that I had a simular problem in my project.

**Comment by mbroadst** [comment]
@jeffmcaffer any chance you have the amqp10 logs (`DEBUG=amqp*` in the environment variables) ?

I'm actually in the process of writing some unit tests around all of this, and am discovering the behavior to be that all `SenderLink` send promises are unconditionally rejected upon reconnect/reattach - a bug indeed, but also would indicate the cases you guys are encountering are almost impossible to reach. Specifically it looks like this:

* Disconnected event occurs, shouldReconnect? then [clearConnectionState](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/amqp_client.js#L181)
* [](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/amqp_client.js#L372)
* if we have a session, then [resetLinkState](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/amqp_client.js#L379)
* [Session.prototype._resetLinkState](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/session.js#L473) basically just calls `forceDetach` on all receiver and sender links in the session
* `SenderLink` will [dispatchPendingSends](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/sender_link.js#L289) for all pending sends with an error, which will call reject on all of them.

So it's rather important to figure out exactly how your programs are getting into this state from tracing the logs, because it's not immediately obvious how this is even possible.

Another unrelated bug seems to be that the Session will [resetLinkState](https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/session.js#L188) on the `Connection.Disconnected` event, whether or not we are reconnecting.

**Comment by jeffmcaffer** [comment]
@mbroadst very interesting.  Thanks for looking at this.  We are definitely, from time to time, seeing the Force Detach errors BUT, I only started seeing them yesterday.  That is, we were having all these problems without ever seeing the amqp:link:detach-forced messages.

One thing that I noticed is that if you call send() when !canSend() then sendMessage is not immediately called, it is added to the pendingSends list and a promise returned.  Note that deferredSender() is not added to the [unsettledSends list](
https://github.com/noodlefrenzy/node-amqp10/blob/master/lib/sender_link.js#L136). So in the flow you described above, forceDetach() does not have a deferredSender to call and settled the promise that was previously returned.

I'll try the repro mentioned in my previous comment and see if messing with canSend() affects this.

Thanks again @mbroadst and @geoffreysamper.  Super helpful to have you engaged on this.

**Comment by geoffreysamper** [comment]
@jeffmcaffer I will first try to simulate with the amqp lib itself. So that I can verify if the issue is in the amqp lib or how amqp lib is used in the project. I had a simular issue in my another project and it was that after a couple of minutes I lost my connection and couldn't recover from it.  
After some analysis we saw that the memory was always increasing and we lost the connection and bam not possible to recover. 
It will be for tomorrow morning I'm living in the utc + 1 timezone. I will try to simulate is as it is used in the project

**Comment by jeffmcaffer** [comment]
Thank @geoffreysamper.  I played around a bit this evening.  Similar to you, I started with just the AMQP code.  To illustrate my scenario (lost promises) I imagined a scenario where the code send()s while !canSend(). Then eventually canSend and _dispatchPendingSends is called.  That triggers the actual send which throws.  In the code below `sender` is a configured and connected SenderLink.  Indeed when it is run, the throw comes all the way out and is not catch by any Promise or AMQP infrastructure.  Neither of the then() args are called.  This may be too artificial but it does illustrate what I think is one of the issues at hand.

```
  const answers = [false, true];
  sender.canSend = () => {
    return answers.shift();
  };
  sender._sendMessage = () => {
    throw new Error('test');
  };

  sender.send({ test: 'message' }).then(
    () => console.log('pushed'),
    error => console.log('failed'));
});
```

**Comment by geoffreysamper** [comment]
@jeffmcaffer Here is a small sample of my [app](https://github.com/geoffreysamper/amqp10-issue-with-reconnection). I will run this in azure to see if I can simulate the problem. Currently I ran it for an hour and I multiple time forced the connection to be killed with cports (windows) or  tcpkill linux and the connection still recovers and I it send messages. Can you take a look at the sample so that we are on the same line. In the one hour that it ran the connection didn't get killed from azure itself. Can you confirm  that you see a lot of reconnections

**Comment by geoffreysamper** [comment]
@jeffmcaffer The sample app  was running the since 10am to 16pm utc+1. I only received on timeout and it was a timeout. I try again in azure

**Comment by jeffmcaffer** [comment]
The sample is roughly inline with what is going on but does not load the queues enough IMHO. There is only one "loop" and only 100 messages per second with a 10ms gap between them.  In our service there are 10+ loops each sending 100+ messages quickly (continuous with 10 concurrent).  The key thing here (for the lost promise case) is that we have to get into a !canSend() condition and then have send errors.  Need to either pound the service or induce issues.

I've been looking at @mbroadst tests and sorta see how to set this up but am not 100% sure how to make the last test (`should send message after reattach`) pass, it currently fails.  My real goal is to then change it to cover the "send message fail causes the promise to be lost" case.

**Comment by mbroadst** [comment]
@jeffmcaffer your pseudo-code above looks pretty good. The last case I added is incomplete because it lacks the second round-trip of connections. I'll try to post an updated case tonight.. I think we've identified a _few_ bugs here imho, so I'll try to carve out some serious time towards the end of the week + weekend for all of this. Thanks for your patience and help both of you.

**Comment by jeffmcaffer** [comment]
@geoffreysamper I [forked your sample repo](https://github.com/jeffmcaffer/amqp10-issue-with-reconnection) to tweak the code to be more aggressive.  @willbar (works with me) will also be poking at that and [my fork of this repo](https://github.com/jeffmcaffer/node-amqp10) to see if we can break it repeatably.

**Comment by geoffreysamper** [comment]
@jeffmcaffer I check you sample and if I understand it correctly code below you want to proof that we lost promises? If this is the case it normal that the completed are less then the attempts because there are still some message that waiting for settelement. If you would replace the  `attempted` with the `completed` you will see that all the message are delivered. The only problem then is that you will remark that the attempt is higher then 1000 and this is because of that we already started a new iteration of sending messages for this loop. I tried to
```
if (messageCount[loop].attempted % 1000 === 0){
     console.log(`attempted 1.000 message, completed ${messageCount[loop].completed} to the queue in loop ${loop}`);
 }
```

I rework my sample that sents 1000 message for each iteration and I wait before all the message of all iteration (loop 0,1,2,3,4,5,6,7,8,9) and then it restart see . If I'm running this you will see that no promises are lost.  

For the other case with reconnection. I looked at the code for reconnection and it make use of the Fibonacci to calcultate the times when it need to reconnect. I will investigated this futher how the program reacts when a connection is killed multiple times [see my repo](https://github.com/geoffreysamper/amqp10-issue-with-reconnection)

**Comment by geoffreysamper** [comment]
@jeffmcaffer @mbroadst Hi, I was able to reproduce the problem for the hangs. Basicly here what happens it related to the reconnections I don't see any promise get lost. 
- connection opened
- connection killed
- reconnect
- depending the number of connection lost the reconnection time will be longer. These are the reconnection times (0,1,1,2,4,8,13,21,34) in seconds. So this mean that the first time the connection will be lost it will  reconnected with a timeout of 0 but when the connection is killed for the ninth time the wait will be 34 seconds before the connection is recovered and during this time there will be pending sends that will lead to long wait times and  it seems like the application is hanging. You can see everything in my repo. I forked the amqp10 lib and add some extra events.

Here is a log that I annotated look at the higher number 21 seconds. 

The only thing is some messages will get the force dettached exception and some will be added to the pendingmessages if we can't send. **take a look at the timestamps**

@mbroadst maybe it can be an idea when the reconnection succeeded that we reset the timeouts.

```
[11:57:05.522] STARTED interval set to 100
[11:57:06.037] connection opened pending message: 0, rejected messages: 0
[11:57:06.057] connected pending message: 0
[11:57:16.501] connection closed pending message: 100
[11:57:16.501] In 0 wil reconnect and num pending messages: 100

[11:57:17.288] connection opened pending message: 800, rejected messages: 159
[11:57:17.305] connected pending message: 800
[11:57:41.266] connection closed pending message: 100
[11:57:41.267] In 1000 wil reconnect and num pending messages: 100

[11:57:42.781] connection opened pending message: 500, rejected messages: 1066
[11:57:42.797] connected pending message: 500
[11:57:48.394] connection closed pending message: 1400
[11:57:48.394] In 1000 wil reconnect and num pending messages: 1400

[11:57:49.764] connection opened pending message: 400, rejected messages: 2699
[11:57:49.780] connected pending message: 400
[11:58:02.089] connection closed pending message: 1800
[11:58:02.090] In 2000 wil reconnect and num pending messages: 1800

[11:58:05.008] connection opened pending message: 900, rejected messages: 4012
[11:58:05.023] connected pending message: 900
[11:58:11.806] connection closed pending message: 1900
[11:58:11.806] In 3000 wil reconnect and num pending messages: 1900

[11:58:15.212] connection opened pending message: 400, rejected messages: 5015
[11:58:15.228] connected pending message: 400
[11:58:21.405] connection closed pending message: 1700
[11:58:21.405] In 5000 wil reconnect and num pending messages: 1700

[HANGS] for almost between 4 and 5 seconds.

[11:58:26.853] connection opened pending message: 400, rejected messages: 6820
[11:58:26.868] connected pending message: 400
[11:58:44.108] connection closed pending message: 100
[11:58:44.108] In 8000 wil reconnect and num pending messages: 100
[HANGS almost 8 seconds]

[11:58:52.576] connection opened pending message: 410, rejected messages: 8032
[11:58:52.594] connected pending message: 480
[11:59:01.245] connection closed pending message: 100
[11:59:01.245] In 13000 wil reconnect and num pending messages: 100

[HANGS] for 13 seconds

[11:59:14.700] connection opened pending message: 400, rejected messages: 13001
[11:59:14.717] connected pending message: 400
[11:59:39.463] connection closed pending message: 100
[11:59:39.463] In 21000 wil reconnect and num pending messages: 100

[HANGS] for almost  21 seconds

[12:00:01.046] connection opened pending message: 400, rejected messages: 21021
[12:00:01.061] connected pending message: 400
[12:00:32.554] connection closed pending message: 100
[12:00:32.555] In 34000 wil reconnect and num pending messages: 100

[HANGS] for almost 34 seconds

[12:01:07.477] connection opened pending message: 510, rejected messages: 33946
[12:01:07.499] connected pending message: 510
[12:04:36.006] connection closed pending message: 70
```

**Comment by jeffmcaffer** [comment]
@geoffreysamper Nice.  Resetting the timeout sequence on reattach seems like an interesting idea but I don't think it would address our hangs.  Ours are permanent hangs (e.g., stops at 0200 and is still hung at 0800).  One related thing to note, most/many/all the hangs we have seen, when the send() is done, the link is in state === 'attaching'. If I understand the code, that means we are actually trying to attach so the timeout hanging is not at play there.

On the repro code I made, the attempted and completed were not expected to match, for the reasons you mentioned.  What should happen though is that the lag should not get particularly big.  Not the greatest indicator.

**Comment by geoffreysamper** [comment]
Do you know their was any cpu usage or did you check the memory usage of the node process.

**Comment by geoffreysamper** [comment]
Hi @jeffmcaffer @mbroadst today I runned my program on azure and set the interval time to 10ms. and serveral time a kept killing the connection via tcpkill I was able to simulate a hang of two minutes nothing more, cpu was hanging on 99% no more logging during this time but eventually it recovered. What I could do setup the ospcrawler and run it on a machine in azure with debug logs enabled. 
 @jeffmcaffer you said something from a rate limitter I would remove this in my test is this a lot of work. 

I know this will take some time to setup but currently for me it the only way to simulate the problem. What do you  both think?

**Comment by mbroadst** [comment]
@jeffmcaffer @geoffreysamper  Hey guys. So I've been playing around with the code today trying to break it, and in the process of refactoring some of the internals of the base `Link` module found an _existing_ test that seemed to reproduce your findings - albeit in a very flakey fashion.

The base of what was going on was a few things:
  - we were using a (imho) superfluous `DETACHING` state, which was probably my fault due to a lack of adherence to the state machine (I'm now going to wade through and fix these issues)
  - the check for whether a connection `shouldAttemptReattach` was happening _before_ the session fully unmapped, but also _in the process_ of unmapping. This is a big no-no, since it will immediately kick off an attempt to reconnect (timeout of 0 when timeouts are generated), even though the rug is about to be pulled out from under it. 

I've pushed the initial fixes to the [issue295](https://github.com/noodlefrenzy/node-amqp10/tree/issue295) branch, and will continue my refactoring there for the time being, until we can resolve all of your issues.

I want to thank you both for your patience on this issue, these are important edge cases we hadn't run into previously and I greatly value the feedback.

**Comment by jeffmcaffer** [comment]
@mbroadst Fantastic news.  Thanks Matt.  If you have something that you want some mileage on, let me and @willbar know.  We have a test setup that we can use to beat on it.  Right now in production are running with my fork but I'm fine moving that to something new pretty aggressively as well.

**Comment by mbroadst** [comment]
@jeffmcaffer go ahead and test using the branch, it passes all our unit/integration tests so afaict it hasn't broken anything in a huge way

**Comment by geoffreysamper** [comment]
So no need to setup the project of of @jeffmcaffer. If I can do anything let me know.

**Comment by mbroadst** [comment]
@geoffreysamper @jeffmcaffer it should be noted that this fix does not attempt to deal with any of the issues related to forced deatchment during reattach, or anything thus related. It simply attempts to fix the initial issue brought up here, specifically that there is an attempt to write to an `undefined` transport/socket.

**Comment by geoffreysamper** [comment]
Hi @mbroadst do I understand it correctly that this fix will eventually not fix the hangs that @jeffmcaffer sees in his project

**Comment by mbroadst** [comment]
@geoffreysamper its incredibly difficult to tell without seeing the actual code and reproducing it on my side, but yeah the intention is to solve those situations as well :)

**Comment by jeffmcaffer** [comment]
@mbroadst That makes sense.  I have been running with your branch and am able to hit the force detach exception pretty much every time. [detach.txt](https://github.com/noodlefrenzy/node-amqp10/files/819912/detach.txt) is a log of a full run with link and link:sender debugging on.  Note that I changed the message logging on send to log the message length rather than the buffer.  Can't release the content.

This example nets out to a simple express server with a route that queues messages (I have not factored the code that way but that's all of consequence that's running).  Incoming requests to queue have batches of up to 10 messages.  The 10 are queued in parallel on the service.  The caller waits for each batch to complete before sending the next.  I start with a newly launched service and an empty Service Bus queue (with partitioning).  Despite the log showing multiple queues, only the "later" queue is in use. Pretty much every time, after about 870 messages it fails as shown in the log.  My monitoring shows that the queue rate is about 20-25/sec (so not much).

Questions

* The entire client is being detached, not just the particular link.  I moved to one client for all four queues but this seems to indicate that I should move back to one client per queue.  What is the recommendation here?
* The forceDetach exception is coming out of the API and causing the request to queue to fail.  I can see how from the library point of view that may make sense but for the caller it makes life very complicated.  Since the client already has a pending sends and retry reconnect mechanism, it would be great if the caller was shielded from this until it is abundantly clear that reconnect is not going to happen (e.g. after a full retry timeout cycle)

FWIW, here is [the code being used to setup the queue](https://github.com/Microsoft/ospo-ghcrawler/blob/develop/lib/amqp10Queue.js).  and here are [the options values.](https://github.com/Microsoft/ospo-ghcrawler/blob/develop/lib/ospoCrawler.js#L78) though none should be at play much here.

**Comment by jeffmcaffer** [comment]
Update.  I tried more configurations.  5 message batches, put in a client side delay to slow things down, factored out the other inactive queues, ... I even went down to 1 message per batch.  All with the same result.  Force detach at 875 messages queued.

OK, so then I removed most of the content from the messages and that let it run longer (did not let it fail). Then I bumped up the content to 80K each and the detach happens after 306 messages.  So, when the server decides to detach appears to be size related.  For the record, 875 of the real messages is 5,725,642 bytes.  306 of the 80K + stuff messages is 24,538,321.  So it is not the absolute size.

No idea what’s going on there but suspect it is a server-ish thing.  Detaches will happen regardless and they need to be handled.

Looking more at the code and the stack trace, it seems that when a reconnection timeout happens, the [pending sends are being cleared](https://github.com/noodlefrenzy/node-amqp10/blob/issue295/lib/sender_link.js#L290) Calling with the error, while !canSend will force sendMessage(error) to be called.  That will [reject the pending promise](https://github.com/noodlefrenzy/node-amqp10/blob/issue295/lib/sender_link.js#L123). 

In the context of a [reconnect attempt](https://github.com/noodlefrenzy/node-amqp10/blob/issue295/lib/amqp_client.js#L189) this seems harsh (I don't know the spec much).  If we are !canSend and call send() then the request is put on the pending list for when we do connect.  Here it seems that if we fail to reconnect once then everything is flushed from the pendingSends list (by virtue of calling sendMessage() with an error.

**Comment by jeffmcaffer** [comment]
OK. one more update.  I put an error handler on the send() call such that if it is a force detach, the caller waits 1 sec and then re-calls send().  Interestingly that worked but not quite how I thought it would.  The reattach etc happens immediately but then there is an 18 second delay between attaching and the sending resuming.

See the log file 
[reattach.txt](https://github.com/noodlefrenzy/node-amqp10/files/820017/reattach.txt)

Would be great if the AMQP client took care of this in a more robust (and timely?) fashion.

**Comment by jeffmcaffer** [comment]
Things are better now with the resend.  Still get forced detach quite frequently (15sec, 2 min, ...) but I've been counting the messages sent for queuing and the ones actually getting queued and they match so the retry on the caller side is working.  

I did however hit this exception and it seems related to this issue.  Perhaps not all the cases are handled as yet.  Seems like the stream was pulled out from under the link's connection.

```
TypeError: Cannot read property 'write' of null
at Object.frames.writeFrame (c:\git\ospo-ghcrawler\node_modules\amqp10\lib\frames.js:64:9)
at Connection.sendFrame (c:\git\ospo-ghcrawler\node_modules\amqp10\lib\connection.js:328:10)
at SenderLink._sendMessage (c:\git\ospo-ghcrawler\node_modules\amqp10\lib\sender_link.js:210:29)
at sendMessage (c:\git\ospo-ghcrawler\node_modules\amqp10\lib\sender_link.js:127:28)
at c:\git\ospo-ghcrawler\node_modules\amqp10\lib\sender_link.js:150:5
at Promise.module.exports.Promise._execute (c:\git\ospo-ghcrawler\node_modules\bluebird\js\release\debuggability.js:300:9)
at Promise.module.exports.Promise._resolveFromExecutor (c:\git\ospo-ghcrawler\node_modules\bluebird\js\release\promise.js:481:18)
at new Promise (c:\git\ospo-ghcrawler\node_modules\bluebird\js\release\promise.js:77:14)
at SenderLink.send (c:\git\ospo-ghcrawler\node_modules\amqp10\lib\sender_link.js:120:10)
at __dirname.push.Q.all.requests.map.request (c:\git\ospo-ghcrawler\lib\amqp10Queue.js:126:26)
at c:\git\ospo-ghcrawler\node_modules\qlimit\index.js:49:26
at _fulfilled (c:\git\ospo-ghcrawler\node_modules\q\q.js:834:54)
at Promise.then.Q.nextTick.self.promiseDispatch.done (c:\git\ospo-ghcrawler\node_modules\q\q.js:863:30)
at Promise.e.Promise.promise.promiseDispatch (c:\git\ospo-ghcrawler\node_modules\q\q.js:796:13)
at c:\git\ospo-ghcrawler\node_modules\q\q.js:857:14
at runSingle (c:\git\ospo-ghcrawler\node_modules\q\q.js:137:13)
```

**Comment by geoffreysamper** [comment]
@jeffmcaffer question for me it weird that the connection is killed so many times. In my case the only time that I get a disconnection is when I not sending message a it timeouts this is by default 1min. Could you have any Idea how many disconnection get in an one hour

**Comment by mbroadst** [comment]
@jeffmcaffer so actually what I'm seeing here is that the initial issue seems to be solved using the new branch (you are not running into cases where you are getting exceptions about attempting to `write` on a null object). Can you confirm this?

The next issues to tackle are as follows:
   * The reason you are getting the "force detach" error is because we drain the pending messages on a force detachment. I brought this up as an issue above, but I think its maybe time to open a new issue for this.
   * We need to figure out why you're getting disconnected so often. This sound Azure related, there is definitely an increase in EH/SB error reports going on presently

**Comment by jeffmcaffer** [comment]
@mbroadst  Unfortunately, the "write on null" issue still occurs.  My [previous comment](https://github.com/noodlefrenzy/node-amqp10/issues/295#issuecomment-284313533) shows a trace of that from last night.  That was only queuing a few thousand messages at 10-20 per second.  They were larger messages but still only on average 6-10K (with a few >200k).  

I'm working around the force detach issue by just retrying the send in my caller. Less than optimal but seems to work.  It does seem that there are some SB issues around the detaching.

**Comment by mbroadst** [comment]
@jeffmcaffer ah sorry I missed that in the flurry of new comments :) bummer. I'll continue my state machine work, I think it'll help things out in a big way.  Do you have any ability to query _why_ this is happening, from the Azure side? I didn't see the actual cause of the detaches in your posted logs above.

**Comment by mbroadst** [comment]
@geoffreysamper @jeffmcaffer quick update

I spent some time tonight trying to track down the cause of #299 , and in the process released v3.5.0 (which is a minor release including removal of the `node-amqp-encoder` and rolling it into an module-local exposed type system.. see gitter for discussion here), and merges the `issue295` code into master.  I'm working with people at Microsoft to identify what happened on the server side for that issue, so we're waiting to hear more on that front.  In the meantime, I overhauled some internals of the `MockServer` and created a very thoroughly tested base test case for reattachment in the `SenderLink` unit tests - I'm hoping this can serve as the foundation for unit testing some of the more eccentric issues we're finding in your individual use cases.
